design/cookie-rewrite-design.md

```diff
@@ -0,0 +1,196 @@
+  - [Nginx](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cookie_domain)
+  - [Apache HTTPD](http://httpd.apache.org/docs/2.2/mod/mod_proxy.html#proxypassreversecookiedomain)
+- Address various open issues related to cookie attributes
+  - https://github.com/projectcontour/contour/issues/2965
+  - https://github.com/projectcontour/contour/issues/3142
+
+## Non Goals
+- Provide full ability to configure Contour/Envoy with arbitrary Lua code
+- Allow users to customize all aspects of an HTTP cookie (value, other attributes not mentioned below, etc.)
+
+## High-Level Design
+`HTTPProxy` will have new fields to configure cookie rewrite policy.
+Cookie attributes related to infrastructure and security considerations will be customizable.
+Other fields, cookie names, and values will not be made customizable.
+The fields we will offer as customizable are as follows:
+- Path
+- Domain
+- Secure
+- SameSite
+
+Initially, the added field for cookie rewriting will be added to the `Route` and `Service` elements of the `HTTPProxy`.
+Envoy provides configuration of Lua using the `typed_per_filter_config` on routes, virtual hosts, or weighted clusters.
+See [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#per-route-configuration) for more.
+
+Additionally, a future step may be to allow customization of the Contour/Envoy generated session affinity cookie directly in the "cookie" load balancing strategy.
+This design will initially require users to customize that via the general setting on the `HTTPProxy` `Route` element.
+We can possibly add some cookie manipulation parameters to the "cookie" load balancing strategy settings so the `X-Contour-Session-Affinity` cookie can be customized more cleanly.
+If/when we move/add the cookie-based hash to the request hash load balancing strategy we provide, we can provide cookie manipulation parameters there to customize any cookies generated by Contour/Envoy.
+See [this draft PR](https://github.com/projectcontour/contour/pull/3993) for an example of how we can move cookie-hashing to the request hash load balancing strategy.
+We can include cookie manipulation parameters in the new `CookieHashOptions` struct.
+
+## Detailed Design
+
+### HTTPProxy API Changes
+
+Below are the changes to the `HTTPProxy` `Route` and `Service` elements and accompanying types:
+
+```golang
+type Route struct {
+  ...
+
+  // +optional
+  CookieRewritePolicy []CookieRewritePolicy `json:"cookieRewritePolicy,omitempty"`
+}
+
+type Service struct {
+  ...
+
+  // +optional
+  CookieRewritePolicy []CookieRewritePolicy `json:"cookieRewritePolicy,omitempty"`
+}
+
+type CookieRewritePolicy struct {
+	// Name is the name of the cookie for which attributes will be rewritten.
+	Name string `json:"name"`
+
+	// +optional
+	PathRewrite *CookiePathRewrite `json:"pathRewrite,omitempty"`
+
+	// +optional
+	DomainRewrite *CookieDomainRewrite `json:"domainRewrite,omitempty"`
```

Comment:
Are these two implemented as structs rather than just string pointers because there may be additional fields added within them over time?

Reply:
yes! in case we want to do regex replace, or some sort of conditional rewrite